File: app/src/main/java/com/byd/dashcast/ui/main/AppActionSheet.java
package com.byd.dashcast.ui.main;

import android.content.Context;
import android.content.Intent;
import android.view.LayoutInflater;
import android.view.View;
import android.widget.ImageView;
import android.widget.TextView;

import com.byd.dashcast.AppLogger;
import com.byd.dashcast.cluster.ClusterService;
import com.byd.dashcast.R;
import com.byd.dashcast.cluster.dpi.ClusterDpiPrefs;
import com.byd.dashcast.cluster.dpi.ClusterResizeActivity;
import com.byd.dashcast.model.AppInfo;

/**
 * Shows the per-app bottom sheet and DPI override dialogs.
 * Stateless — call {@link #show(AppInfo, Host)} from the host activity.
 */
public final class AppActionSheet {

    private static final String TAG = "AppActionSheet";

    private AppActionSheet() {}

    public interface Host {
        Context getContext();
        boolean isActivityAlive();
        void startActivity(Intent intent);
        ClusterService getClusterServiceIfBound();
        /** Package currently projected on the cluster. */
        String getCurrentClusterPkg();
        /** Package last sent to the main display. */
        String getMainDisplayPkg();
        void onSendToDashboard(AppInfo app);
        void onSendToMain(AppInfo app);
        void onKillApp(AppInfo app);
        void onToggleFavorite(AppInfo app);
        void onSetAutoLaunch(AppInfo app, boolean enable);
    }

    // ── Public entry point ────────────────────────────────────────────────────

    public static void show(final AppInfo app, final Host host) {
        if (app == null || !host.isActivityAlive()) return;

        Context ctx = host.getContext();
        final com.google.android.material.bottomsheet.BottomSheetDialog dialog =
                new com.google.android.material.bottomsheet.BottomSheetDialog(ctx);
        View v = LayoutInflater.from(ctx).inflate(R.layout.dialog_app_actions, null);
        dialog.setContentView(v);

        ImageView icon     = v.findViewById(R.id.sheet_icon);
        TextView  name     = v.findViewById(R.id.sheet_name);
        TextView  pkg      = v.findViewById(R.id.sheet_pkg);
        com.google.android.material.materialswitch.MaterialSwitch swAuto =
                v.findViewById(R.id.sheet_sw_auto);
        View      rowFav   = v.findViewById(R.id.sheet_action_favorite);
        TextView  lblFav   = v.findViewById(R.id.sheet_lbl_favorite);
        View      rowToMain  = v.findViewById(R.id.sheet_action_to_main);
        View      rowToClus  = v.findViewById(R.id.sheet_action_to_cluster);
        View      rowResize  = v.findViewById(R.id.sheet_action_resize);
        View      rowDpi     = v.findViewById(R.id.sheet_action_dpi);
        TextView  tvDpiVal   = v.findViewById(R.id.sheet_dpi_value);
        View      rowKill    = v.findViewById(R.id.sheet_action_kill);

        icon.setImageDrawable(app.icon);
        name.setText(app.appName);
        pkg.setText(app.packageName);
        lblFav.setText(app.isFavorite
                ? R.string.sheet_remove_favorite
                : R.string.sheet_add_favorite);

        swAuto.setChecked(app.isAutoLaunch);
        swAuto.setOnCheckedChangeListener((b, checked) -> host.onSetAutoLaunch(app, checked));

        rowFav.setOnClickListener(vv -> {
            host.onToggleFavorite(app);
            dialog.dismiss();
        });

        final boolean isActive = app.packageName != null
                && app.packageName.equals(host.getCurrentClusterPkg());
        final boolean isOnMain = app.packageName != null
                && app.packageName.equals(host.getMainDisplayPkg());
        rowToMain.setVisibility(isActive ? View.VISIBLE : View.GONE);
        rowToClus.setVisibility(isOnMain || (!isActive && !isOnMain) ? View.VISIBLE : View.GONE);
        rowResize.setVisibility(isActive ? View.VISIBLE : View.GONE);
        rowKill.setVisibility((isActive || isOnMain) ? View.VISIBLE : View.GONE);

        rowToMain.setOnClickListener(vv -> { host.onSendToMain(app); dialog.dismiss(); });
        rowToClus.setOnClickListener(vv -> { host.onSendToDashboard(app); dialog.dismiss(); });

        rowResize.setOnClickListener(vv -> {
            ClusterService svc = host.getClusterServiceIfBound();
            int displayId = svc != null ? svc.getDisplayId() : 1;
            if (displayId < 0) displayId = 1;
            Intent it = new Intent(ctx, ClusterResizeActivity.class);
            it.putExtra(ClusterResizeActivity.EXTRA_PACKAGE, app.packageName);
            it.putExtra(ClusterResizeActivity.EXTRA_DISPLAY_ID, displayId);
            host.startActivity(it);
            dialog.dismiss();
        });

        final int curDpi = ClusterDpiPrefs.getDpi(ctx, app.packageName);
        if (curDpi > 0) {
            tvDpiVal.setText(ctx.getString(R.string.sheet_dpi_value_fmt, curDpi));
        } else {
            tvDpiVal.setText(R.string.sheet_dpi_value_default);
        }
        rowDpi.setOnClickListener(vv -> {
            dialog.dismiss();
            showDpiDialog(app, host);
        });

        rowKill.setOnClickListener(vv -> {
            host.onKillApp(app);
            dialog.dismiss();
        });

        dialog.show();
        // v0.9.73 — open fully expanded (skip the half-collapsed peek that hides actions).
        try {
            com.google.android.material.bottomsheet.BottomSheetBehavior<?> b = dialog.getBehavior();
            b.setSkipCollapsed(true);
            b.setState(com.google.android.material.bottomsheet.BottomSheetBehavior.STATE_EXPANDED);
        } catch (Throwable t) {
            AppLogger.w(TAG, "BottomSheet expand failed: " + t.getMessage());
        }
    }

    // ── DPI dialogs ───────────────────────────────────────────────────────────

    // v1.2.81 — Cluster per-app DPI override picker.
    // Presets cover the common cluster-friendly densities; "Custom" accepts 96..480 dpi.
    // Applied at next launch on the cluster display via ClusterDpiManager.applyForLaunch.
    @android.annotation.SuppressLint("InflateParams")
    static void showDpiDialog(final AppInfo app, final Host host) {
        if (app == null || app.packageName == null || !host.isActivityAlive()) return;

        Context ctx = host.getContext();
        final int[] presets = { 120, 160, 200, 240, 280, 320 };
        final int current = ClusterDpiPrefs.getDpi(ctx, app.packageName);

        final String[] labels = new String[presets.length + 2];
        labels[0] = ctx.getString(R.string.dpi_default);
        for (int i = 0; i < presets.length; i++) {
            labels[i + 1] = ctx.getString(R.string.dpi_value_fmt, presets[i]);
        }
        labels[labels.length - 1] = ctx.getString(R.string.dpi_custom);

        int checked = 0;
        if (current > 0) {
            checked = labels.length - 1;
            for (int i = 0; i < presets.length; i++) {
                if (presets[i] == current) { checked = i + 1; break; }
            }
        }

        new androidx.appcompat.app.AlertDialog.Builder(ctx)
                .setTitle(ctx.getString(R.string.dpi_dialog_title, app.appName))
                .setSingleChoiceItems(labels, checked, (d, which) -> {
                    if (which == 0) {
                        ClusterDpiPrefs.setDpi(ctx, app.packageName, 0);
                        d.dismiss();
                        android.widget.Toast.makeText(ctx,
                                R.string.dpi_applied_next_launch,
                                android.widget.Toast.LENGTH_SHORT).show();
                    } else if (which == labels.length - 1) {
                        d.dismiss();
                        showDpiCustomDialog(app, current, host);
                    } else {
                        ClusterDpiPrefs.setDpi(ctx, app.packageName, presets[which - 1]);
                        d.dismiss();
                        android.widget.Toast.makeText(ctx,
                                R.string.dpi_applied_next_launch,
                                android.widget.Toast.LENGTH_SHORT).show();
                    }
                })
                .setNegativeButton(android.R.string.cancel, null)
                .show();
    }

    @android.annotation.SuppressLint("InflateParams")
    private static void showDpiCustomDialog(final AppInfo app, final int prefill,
                                            final Host host) {
        Context ctx = host.getContext();
        final android.widget.EditText input = new android.widget.EditText(ctx);
        input.setInputType(android.text.InputType.TYPE_CLASS_NUMBER);
        input.setHint(ctx.getString(R.string.dpi_custom_hint,
                ClusterDpiPrefs.MIN_DPI, ClusterDpiPrefs.MAX_DPI));
        if (prefill > 0) input.setText(String.valueOf(prefill));

        int pad = (int) (16 * ctx.getResources().getDisplayMetrics().density);
        android.widget.FrameLayout wrap = new android.widget.FrameLayout(ctx);
        wrap.setPadding(pad, pad / 2, pad, 0);
        wrap.addView(input);

        new androidx.appcompat.app.AlertDialog.Builder(ctx)
                .setTitle(R.string.dpi_custom)
                .setView(wrap)
                .setPositiveButton(android.R.string.ok, (d, w) -> {
                    int dpi;
                    try { dpi = Integer.parseInt(input.getText().toString().trim()); }
                    catch (NumberFormatException e) { dpi = 0; }
                    if (dpi < ClusterDpiPrefs.MIN_DPI || dpi > ClusterDpiPrefs.MAX_DPI) {
                        android.widget.Toast.makeText(ctx,
                                ctx.getString(R.string.dpi_custom_hint,
                                        ClusterDpiPrefs.MIN_DPI, ClusterDpiPrefs.MAX_DPI),
                                android.widget.Toast.LENGTH_SHORT).show();
                        return;
                    }
                    ClusterDpiPrefs.setDpi(ctx, app.packageName, dpi);
                    android.widget.Toast.makeText(ctx,
                            R.string.dpi_applied_next_launch,
                            android.widget.Toast.LENGTH_SHORT).show();
                })
                .setNegativeButton(android.R.string.cancel, null)
                .show();
    }
}

File: app/src/main/java/com/byd/dashcast/ui/main/AppActionSheet.kt
package com.byd.dashcast.ui.main

import android.annotation.SuppressLint
import android.content.Context
import android.content.Intent
import android.text.InputType
import android.view.LayoutInflater
import android.view.View
import android.widget.EditText
import android.widget.FrameLayout
import android.widget.ImageView
import android.widget.TextView
import android.widget.Toast
import androidx.appcompat.app.AlertDialog
import com.byd.dashcast.R
import com.byd.dashcast.cluster.ClusterService
import com.byd.dashcast.cluster.dpi.ClusterDpiPrefs
import com.byd.dashcast.cluster.dpi.ClusterResizeActivity
import com.byd.dashcast.data.prefs.ClusterPrefs
import com.byd.dashcast.model.AppInfo
import com.byd.dashcast.ui.settings.SettingsActivity
import com.byd.dashcast.util.AppLogger
import com.google.android.material.bottomsheet.BottomSheetBehavior
import com.google.android.material.bottomsheet.BottomSheetDialog
import com.google.android.material.materialswitch.MaterialSwitch

/**
 * Shows the per-app bottom sheet and DPI override dialogs.
 * Stateless — call [show] from the host activity.
 */
object AppActionSheet {

    private const val TAG = "AppActionSheet"

    interface Host {
        fun getContext(): Context
        fun isActivityAlive(): Boolean
        fun startActivity(intent: Intent)
        fun getClusterServiceIfBound(): ClusterService?
        /** Package currently projected on the cluster. */
        fun getCurrentClusterPkg(): String?
        /** Package last sent to the main display. */
        fun getMainDisplayPkg(): String?
        /** Packages currently projected via a fission layout (headless orchestrator). */
        fun getLayoutPkgs(): Set<String>?
        fun onSendToDashboard(app: AppInfo)
        fun onSendToMain(app: AppInfo)
        fun onKillApp(app: AppInfo)
        fun onToggleFavorite(app: AppInfo)
        fun onSetAutoLaunch(app: AppInfo, enable: Boolean)
    }

    // ── Public entry point ────────────────────────────────────────────────────

    @JvmStatic
    fun show(app: AppInfo?, host: Host) {
        if (app == null || !host.isActivityAlive()) return

        val ctx = host.getContext()
        val dialog = BottomSheetDialog(ctx)
        val v = LayoutInflater.from(ctx).inflate(R.layout.dialog_app_actions, null)
        dialog.setContentView(v)

        val icon = v.findViewById<ImageView>(R.id.sheet_icon)
        val name = v.findViewById<TextView>(R.id.sheet_name)
        val pkg = v.findViewById<TextView>(R.id.sheet_pkg)
        val swAuto = v.findViewById<MaterialSwitch>(R.id.sheet_sw_auto)
        val rowFav = v.findViewById<View>(R.id.sheet_action_favorite)
        val lblFav = v.findViewById<TextView>(R.id.sheet_lbl_favorite)
        val rowToMain = v.findViewById<View>(R.id.sheet_action_to_main)
        val rowToClus = v.findViewById<View>(R.id.sheet_action_to_cluster)
        val rowResize = v.findViewById<View>(R.id.sheet_action_resize)
        val rowDpi = v.findViewById<View>(R.id.sheet_action_dpi)
        val tvDpiVal = v.findViewById<TextView>(R.id.sheet_dpi_value)
        val rowKill = v.findViewById<View>(R.id.sheet_action_kill)

        icon.setImageDrawable(app.icon)
        name.text = app.appName
        pkg.text = app.packageName
        lblFav.setText(if (app.isFavorite) R.string.sheet_remove_favorite else R.string.sheet_add_favorite)

        swAuto.isChecked = app.isAutoLaunch
        swAuto.setOnCheckedChangeListener { _, checked -> host.onSetAutoLaunch(app, checked) }

        rowFav.setOnClickListener {
            host.onToggleFavorite(app)
            dialog.dismiss()
        }

        val isActive = app.packageName == host.getCurrentClusterPkg()
        val isOnMain = app.packageName == host.getMainDisplayPkg()
        // Fission-layout app: projected via the headless orchestrator on a dedicated VD slot.
        // The classic move/resize actions go through ClusterService and don't know about the
        // fission VD, so only expose kill (routed to FissionOrchestrator.killLayoutSlot).
        val layoutPkgs = host.getLayoutPkgs()
        val isLayout = layoutPkgs != null && layoutPkgs.contains(app.packageName)
        rowToMain.visibility = if (isActive) View.VISIBLE else View.GONE
        rowToClus.visibility = if (!isActive && !isLayout) View.VISIBLE else View.GONE
        rowResize.visibility = if (isActive) View.VISIBLE else View.GONE
        rowKill.visibility = if (isActive || isOnMain || isLayout) View.VISIBLE else View.GONE

        rowToMain.setOnClickListener { host.onSendToMain(app); dialog.dismiss() }
        rowToClus.setOnClickListener { host.onSendToDashboard(app); dialog.dismiss() }

        rowResize.setOnClickListener {
            val svc = host.getClusterServiceIfBound()
            var displayId = svc?.getDisplayId() ?: 1
            if (displayId < 0) displayId = 1
            val resizeIntent = Intent(ctx, ClusterResizeActivity::class.java)
            resizeIntent.putExtra(ClusterResizeActivity.EXTRA_PACKAGE, app.packageName)
            resizeIntent.putExtra(ClusterResizeActivity.EXTRA_DISPLAY_ID, displayId)
            // Re-open on the last saved rectangle so the frame reflects the current size
            // instead of resetting to full-screen.
            readSavedClusterRect(ctx, app.packageName)?.let {
                resizeIntent.putExtra(ClusterResizeActivity.EXTRA_INIT_LTRB, it)
            }
            host.startActivity(resizeIntent)
            dialog.dismiss()
        }

        val curDpi = ClusterDpiPrefs.getDpi(ctx, app.packageName)
        if (curDpi > 0) {
            tvDpiVal.text = ctx.getString(R.string.sheet_dpi_value_fmt, curDpi)
        } else {
            tvDpiVal.setText(R.string.sheet_dpi_value_default)
        }
        rowDpi.setOnClickListener {
            dialog.dismiss()
            showDpiDialog(app, host)
        }

        rowKill.setOnClickListener {
            host.onKillApp(app)
            dialog.dismiss()
        }

        dialog.show()
        // v0.9.73 — open fully expanded (skip the half-collapsed peek that hides actions).
        try {
            val b = dialog.behavior
            b.skipCollapsed = true
            b.state = BottomSheetBehavior.STATE_EXPANDED
        } catch (t: Throwable) {
            AppLogger.w(TAG, "BottomSheet expand failed: " + t.message)
        }
    }

    // ── DPI dialogs ───────────────────────────────────────────────────────────

    // v1.2.81 — Cluster per-app DPI override picker. "Custom" accepts MIN..MAX dpi.
    @SuppressLint("InflateParams")
    private fun showDpiDialog(app: AppInfo?, host: Host) {
        if (app == null || !host.isActivityAlive()) return

        val ctx = host.getContext()
        val presets = intArrayOf(120, 160, 200, 240, 280, 320)
        val current = ClusterDpiPrefs.getDpi(ctx, app.packageName)

        val labels = arrayOfNulls<String>(presets.size + 2)
        labels[0] = ctx.getString(R.string.dpi_default)
        for (i in presets.indices) {
            labels[i + 1] = ctx.getString(R.string.dpi_value_fmt, presets[i])
        }
        labels[labels.size - 1] = ctx.getString(R.string.dpi_custom)

        var checked = 0
        if (current > 0) {
            checked = labels.size - 1
            for (i in presets.indices) {
                if (presets[i] == current) {
                    checked = i + 1
                    break
                }
            }
        }

        AlertDialog.Builder(ctx)
            .setTitle(ctx.getString(R.string.dpi_dialog_title, app.appName))
            .setSingleChoiceItems(labels, checked) { d, which ->
                when {
                    which == 0 -> {
                        ClusterDpiPrefs.setDpi(ctx, app.packageName, 0)
                        d.dismiss()
                        Toast.makeText(ctx, R.string.dpi_applied_next_launch, Toast.LENGTH_SHORT).show()
                    }
                    which == labels.size - 1 -> {
                        d.dismiss()
                        showDpiCustomDialog(app, current, host)
                    }
                    else -> {
                        ClusterDpiPrefs.setDpi(ctx, app.packageName, presets[which - 1])
                        d.dismiss()
                        Toast.makeText(ctx, R.string.dpi_applied_next_launch, Toast.LENGTH_SHORT).show()
                    }
                }
            }
            .setNegativeButton(android.R.string.cancel, null)
            .show()
    }

    // SetTextI18n: prefill is a plain integer dpi value (String.valueOf in the
    // original Java) — .toString() trips the stricter Kotlin lint.
    @SuppressLint("InflateParams", "SetTextI18n")
    private fun showDpiCustomDialog(app: AppInfo, prefill: Int, host: Host) {
        val ctx = host.getContext()
        val input = EditText(ctx)
        input.inputType = InputType.TYPE_CLASS_NUMBER
        input.hint = ctx.getString(R.string.dpi_custom_hint, ClusterDpiPrefs.MIN_DPI, ClusterDpiPrefs.MAX_DPI)
        if (prefill > 0) input.setText(prefill.toString())

        val pad = (16 * ctx.resources.displayMetrics.density).toInt()
        val wrap = FrameLayout(ctx)
        wrap.setPadding(pad, pad / 2, pad, 0)
        wrap.addView(input)

        AlertDialog.Builder(ctx)
            .setTitle(R.string.dpi_custom)
            .setView(wrap)
            .setPositiveButton(android.R.string.ok) { _, _ ->
                val dpi = try {
                    input.text.toString().trim().toInt()
                } catch (e: NumberFormatException) {
                    0
                }
                if (dpi < ClusterDpiPrefs.MIN_DPI || dpi > ClusterDpiPrefs.MAX_DPI) {
                    Toast.makeText(
                        ctx,
                        ctx.getString(R.string.dpi_custom_hint, ClusterDpiPrefs.MIN_DPI, ClusterDpiPrefs.MAX_DPI),
                        Toast.LENGTH_SHORT
                    ).show()
                    return@setPositiveButton
                }
                ClusterDpiPrefs.setDpi(ctx, app.packageName, dpi)
                Toast.makeText(ctx, R.string.dpi_applied_next_launch, Toast.LENGTH_SHORT).show()
            }
            .setNegativeButton(android.R.string.cancel, null)
            .show()
    }

    /** Reads the per-app hand-drawn cluster rectangle ("l,t,r,b"), or null if none/invalid. */
    private fun readSavedClusterRect(ctx: Context, pkg: String): IntArray? {
        val s = ctx.getSharedPreferences(ClusterPrefs.PREFS_NAME, Context.MODE_PRIVATE)
            .getString(SettingsActivity.PREF_CLUSTER_RECT_PREFIX + pkg, null) ?: return null
        val parts = s.split(",")
        if (parts.size != 4) return null
        return try {
            intArrayOf(parts[0].toInt(), parts[1].toInt(), parts[2].toInt(), parts[3].toInt())
        } catch (e: NumberFormatException) {
            null
        }
    }
}
